feat(KL-68): update Product API

ohhamma opened the pull request:
## 📌 연관된 이슈
- **[KL-68/상품 정보 수정 API 구현](https://ohhamma.atlassian.net/browse/KL-68)**

## 📝 작업 내용
- 상품 정보 수정 API 구현
- 관련 테스트코드 작성

## 🌳 작업 브랜치명
- `KL-68/상품-정보-수정-api-구현`

## 📸 스크린샷 (선택)

## 💬 리뷰 요구사항 (선택)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Introduced a new endpoint for updating product information with PATCH requests, allowing partial updates to product details.
	- Enhanced product creation and update functionalities with improved validation for incoming requests.

- **Bug Fixes**
	- Improved error handling for malformed requests, enhancing clarity and robustness in error reporting.

- **Tests**
	- Expanded test coverage to include scenarios for product updates and validation, ensuring robustness in functionality.

- **Documentation**
	- Updated validation messages and constraints for product-related data to improve data integrity.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ohhamma committed:
KL-68/feat: update product

- updateDto를 받아 null이 아닌 필드 값을 변경하는 상품 update 메소드 작성
- 관련 테스트코드 작성
  - 상품 정보 업데이트 테스트
  - 상품 정보 부분 업데이트 테스트
  - 상품 정보 업데이트 시 null 값 무시 테스트

ohhamma committed:
KL-68/feat: update product service

- ProductService에 id, updateDto를 받아 상품의 정보를 수정 및 레포지토리에 변경사항 저장
- 관련 테스트코드 작성
  - 상품 업데이트 성공 테스트
  - 상품 업데이트 실패 테스트 - 상품 없음

ohhamma committed:
KL-68/feat: update product api

- 상품 정보 수정 API 구현
- 관련 테스트코드 작성
  - 상품 정보 부분 업데이트 API 테스트

idealflower-k reviewed (changes requested):
상품 속성들의 유효성검사가 추가되면 좋을거 같습니다!

ohhamma committed:
KL-68/fix: validate product create request

- 상품 등록 요청 유효성 검사
- 컨트롤러에 @Valid 키워드 추가
- 관련 테스트코드 추가
  - 상품 등록 API 유효성 검사 실패 테스트
- 유효성 검사에서 발생한 오류 메시지를 수집해서 출력하도록 GlobalExceptionHandler 수정
- 수정된 내용에 맞게 테스트코드 수정
  - BindingResult를 모킹해서 getFieldErrors()가 유효한 FieldError 목록을 반환하도록 설정
- h2 데이터베이스 모드 MySQL로 설정

ohhamma committed:
KL-68/feat: add HttpMessageNotReadableException handler

- 유효하지 않은 요청 메시지에 대한 예외 핸들링
- 관련 테스트코드 추가

ohhamma committed:
KL-68/feat: add product price constraint

- 상품 가격이 0 이상이어야 한다는 제약 추가
- 관련 테스트코드 추가
  - 가격이 음수일 때 유효성 검사 실패
  - 가격이 0일 때 유효성 검사 성공
- 상품 요청 dto에서 가격과 주소 자리 바꿈
- 바뀐 자리에 맞게 관련 테스트코드 수정

ohhamma committed:
KL-68/test: modify GlobalExceptionHandlerTest

- 각 테스트케이스에서 발생한 예외에 대한 코드와 메시지를 ErrorCode의 예외의 값들과 비교하도록 수정
- 지원하지 않는 메소드 예외의 이름을 METHOD_NOT_SUPPORTED로 변경

ohhamma committed:
KL-68/feat: validate product creation

- 상품 등록 요청 dto의 유효성 검사 추가
  - 상품명, 상품 설명, 주소의 길이 검사
  - 필수 String 필드에서 빈 문자열 검사
  - 가격의 범위 검사
- 추가된 검사 내용 테스트코드에 추가
- 상품 등록 요청 dto의 필드 순서 변경
  - 이에 맞게 테스트코드 수정
- 상품 필드의 max값 constants로 관리
  - 이름, 설명, 주소의 각 최대 길이

ohhamma committed:
KL-68/chore: use constants for validation messages

- 상품 등록 유효성 검사 시 보여주는 메시지들을 상수로 관리
- 이에 맞게 테스트코드 수정

ohhamma committed:
KL-68/chore: add final for constant classes

- 상수 관리하는 클래스에 final 키워드 추가

ohhamma committed:
KL-68/feat: validate product update request

- 상품 정보 수정 시 유효성 검사
  - 상품명, 상품설명, 주소 길이 검사
  - 가격 범위 검사
- 상품 정보 수정 dto 필드 순서 변경
  - 이에 따라 테스트코드 수정

idealflower-k commented on src/main/java/taco/klkl/domain/product/controller/ProductController.java:
```diff
@@ -33,8 +36,20 @@ public ResponseEntity<ProductDetailResponseDto> getProductInfoById(@PathVariable
 	@Operation(summary = "상품 등록", description = "상품을 등록합니다.")
-	public ResponseEntity<ProductDetailResponseDto> createProduct(@RequestBody ProductCreateRequestDto createRequest) {
+	public ResponseEntity<ProductDetailResponseDto> createProduct(
+		@Valid @RequestBody ProductCreateRequestDto createRequest
```
`@Valid` 어노테이션은 지금과 같은 상황에서만 사용이 가능한 어노테이션인가요?

ohhamma replied:
`@Valid`는 요청 dto의 제약 조건 어노테이션들을 확인해주는 어노테이션이에요! 잘못된 데이터가 핵심 로직에 접근하기 전에 어플리케이션 레벨에서 유효성 검사해줄 때 사용합니다. 즉, POST나 PUT, PATCH 메소드처럼 요청 body에 데이터가 들어오는 경우에 주로 사용합니다.

idealflower-k commented on src/main/java/taco/klkl/domain/product/domain/Product.java:
```diff
@@ -35,34 +36,62 @@ public class Product {
+		name = "name",
+		length = ProductConstants.NAME_MAX_LENGTH,
+		nullable = false
+	)
```
오 이렇게 작성하는게 가독성이 더 좋네요!

ohhamma replied:
맞아요! 가독성이 더 좋아서 각 필드마다 개행을 넣어줬어요.

idealflower-k commented on src/main/java/taco/klkl/domain/product/service/ProductService.java:
```diff
@@ -7,31 +7,43 @@
-	public ProductDetailResponseDto getProductInfoById(Long id) {
-		Product product = productRepository.findById(id)
+	public ProductDetailResponseDto getProductInfoById(final Long id) {
+		final Product product = productRepository.findById(id)
```
매개변수와 지역변수에도 final 키워드를 사용해주는게 권장되는 방법인건가요?

ohhamma replied:
`final` 키워드를 붙여서 객체의 불변성을 보장하고, 코드 상으로도 해당 객체가 불변일거라는 사실을 줄 수 있어서 넣는 것이 좋다고 생각합니다!

idealflower-k commented on src/main/java/taco/klkl/domain/product/service/ProductService.java:
```diff
@@ -7,31 +7,43 @@
 
 @Service
-@Transactional
+@Transactional(readOnly = true)
```
클래스에는 readOnly를 설정하고 메서드별로 필요에 따라 트랜잭션을 설정해주는 방법이 확실히 안전한거 같습니다!

ohhamma replied:
맞습니다! `@Transactional`의 디폴트 값이 `readOnly = false`이기 때문에, 읽기 쓰기가 모두 가능하더라고요.

min3m replied:
아주 좋은 방법인거 같습니다! 클래스에선 readonly로 걸어두고 update나 create를 하는 매서드 내에만 Transactional을 붙이는거!

idealflower-k commented on src/main/java/taco/klkl/global/common/constants/ProductConstants.java:
```diff
@@ -1,11 +1,15 @@
 package taco.klkl.global.common.constants;
 
-public class ProductConstants {
+public final class ProductConstants {
```
상수를 이렇게 관리하는게 좋은거 같습니다! 제 코드도 리팩토링을 해야겠습니다.

idealflower-k commented on src/test/java/taco/klkl/domain/product/dto/request/ProductCreateRequestDtoTest.java:
```diff
@@ -23,160 +27,236 @@ void setUp() {
+	@ValueSource(ints = {101, 200, 1000})
+	@DisplayName("상품명이 최대 길이를 초과할 때 검증 실패")
+	void productNameExceedsMaxLength(int nameLength) {
+		String longName = "a".repeat(nameLength);
```
`@ParametterizedTest`, `@ValueSource` 이 어노테이션들은 어떤 동작을 해주나요??

ohhamma replied:
`@ParametterizedTest`는 하나의 테스트 메소드를 여러 다른 입력값으로 반복 실행할 수 있게 해줍니다. 이때 메소드에 들어가는 입력값을 `@ValueSource`로 제공을 하고 있는 형태입니다! 현재 메소드에서는 이름의 길이를 테스트 메소드의 argument로 받게끔 구현했습니다.

idealflower-k commented on src/main/java/taco/klkl/global/error/GlobalExceptionHandler.java:
```diff
@@ -73,6 +103,7 @@ public ResponseEntity<Object> handleCustomException(CustomException ex) {
 		return ResponseEntity.status(errorCode.getStatus()).body(globalResponse);
 	}
 
+	// TODO: json 문법 오류 있으면 BAD_REQUEST 반환
```
이제 이 TODO는 없어도 될 거 같습니다!

ohhamma replied:
앗 삭제한다는걸 깜빡했군요! 수정하겠습니다.

ohhamma committed:
KL-68/chore: remove comment

- 필요 없는 TODO 코멘트 삭제

idealflower-k reviewed (dismissed):
상품이 할게 많은데 고생하셨습니다!
역시 팀장님 코딩 천재시네요 :)

ohhamma committed:
KL-68/feat: remove save in update

- update 로직에 불필요한 save 메소드 호출 제거
- 이에 따른 테스트코드 수정

ohhamma commented:
@coderabbitai review

min3m commented on src/main/java/taco/klkl/global/common/constants/ProductValidationMessages.java:
```diff
@@ -0,0 +1,23 @@
+package taco.klkl.global.common.constants;
```
validation message를 enum 대신 불변 클래스로 작성하신 이유가 있나요??

ohhamma replied:
enum을 사용하면 final 클래스를 사용할 때보다 메모리 사용량도 많고, 상수가 많아지면 컴파일 시간도 늘어날 수 있다고 합니다. validation message는 단순 문자열 상수만 담기 때문에 주로 final 클래스를 많이 사용한다고 합니다!

min3m reviewed (approved):
해당 도메인에서 공통적으로 사용되는 값들을 상수로 관리하는 방법은 좋아보입니다!!! 또한 트랜잭션 관리 어떻게 해야하나 고민중이었는데 덕분에 방법을 배워간 것 같습니다~ 고생하셨어요!

idealflower-k reviewed (approved):
JpaRepository가 엔티티 객체의 변경을 감지하고 명시적인 save()호출 없이 db적용가능한 점을 알 수 있었습니다!
고생하셨습니다! :)
